builtin/providers/aws/resource_aws_autoscaling_group.go

```diff
@@ -518,20 +537,34 @@ func resourceAwsAutoscalingGroupDrain(d *schema.ResourceData, meta interface{})
 		return nil
 	}
 
-	// First, set the capacity to zero so the group will drain
-	log.Printf("[DEBUG] Reducing autoscaling group capacity to zero")
-	opts := autoscaling.UpdateAutoScalingGroupInput{
-		AutoScalingGroupName: aws.String(d.Id()),
-		DesiredCapacity:      aws.Int64(0),
-		MinSize:              aws.Int64(0),
-		MaxSize:              aws.Int64(0),
-	}
-	if _, err := conn.UpdateAutoScalingGroup(&opts); err != nil {
-		return fmt.Errorf("Error setting capacity to zero to drain: %s", err)
+	if delay, ok := d.GetOk("delete_delay"); ok {
+		g, err := getAwsAutoscalingGroup(d.Id(), conn)
+		if err != nil {
+			return err
+		}
+		delayT := time.Duration(delay.(int)) * time.Second
+		t := time.Now()
+		count := *g.DesiredCapacity
+		for count > 0 {
+			count--
+			if err := resourceAwsAutoscalingScaleDownToCount(d, count, conn); err != nil {
+				return err
+			}
+			now := time.Now()
```

Comment:
Do we really want to sleep even after the final "scale to 0"?  I think maybe there should be a `if count == 0 { break }` here